### PR TITLE
Expose display option to add ! on non-nullable types

### PR DESCRIPTION
**Pull request by jcouv**
To display `!` in shipped API files (WIP PR at https://github.com/dotnet/roslyn-analyzers/pull/3125), we'll need to expose the previously-internal display option.

### CODE REVIEW

**Comment by sharwell** [comment]
There is no need to expose this flag for PublicApiAnalyzers. Regardless of whether this flag is internal or public, the analyzer will be using the flag only via reflection.

**Comment by jcouv** [comment]
>  the analyzer will be using the flag only via reflection.

What makes you say that? The analyzer for shipped APIs makes a `SymbolDisplayFormat` from the published API, not using reflection as far as I can tell.

https://github.com/dotnet/roslyn-analyzers/blob/master/src/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.cs#L114-L134

**Comment by sharwell** [comment]
@jcouv the analyzer references an old version of the compiler package, so unless the new flag is back-ported to Roslyn 1.x, it would only be accessible via light-up (reflection).

**Comment by jcouv** [comment]
@sharwell Reflection is not the only way to use the new flag. You can cast a number to an enum type. That's what I did in [WIP PR](https://github.com/dotnet/roslyn-analyzers/pull/3125/files#diff-a4e6025e7e8977728a43d7e91bfcea39R136).

``` chsarp
private const int IncludeNullableReferenceTypeModifier = 1 << 6;

...
                    SymbolDisplayMiscellaneousOptions.UseSpecialTypes |
                    (SymbolDisplayMiscellaneousOptions)IncludeNullableReferenceTypeModifier); // cast of constant to enum type
```

This hard-coding is safe because the enum value is part of public API (ie. won't change).

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for review. This PR will unblock an analyzer PR (to include nullability information in shipped APIs). 
The updated analyzer will be able to pass this value even though the analyzer will not take a dependency on new compiler bits, by casting a hard-coded constant to enum type. See [WIP PR](https://github.com/dotnet/roslyn-analyzers/pull/3125/files#diff-a4e6025e7e8977728a43d7e91bfcea39R136) for the analyzer.

**Comment by jcouv** [comment]
@333fred @cston Would you have time to review this simple change? That will unblock another PR I'm working on (in roslyn-analyzers). Thanks

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 1]
```diff
@@ -79,15 +79,16 @@ private bool ShouldAddNullableAnnotation(ITypeSymbol type)
             switch (type.NullableAnnotation)
             {
                 case CodeAnalysis.NullableAnnotation.Annotated:
-                    if (format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier) &&
+                    if ((format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.IncludeNullableReferenceTypeModifier) ||
+                        format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.IncludeNonNullableReferenceTypeModifier)) &&
```
Comment (cston) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs]: Is this change needed? (It seems strange that `IncludeNonNullable...` implies `IncludeNullable...` but not the reverse.)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs]: I'd be okay with making the two flags independent, but I couldn't think of a scenario when we would want to print `!` annotations but not `?` annotations.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs]: @333fred What's your thought on the relationship of these two flags?

Previously, we asserted that we only use the `!` flag when we also use the `?` flag. But an assertion isn't sufficient for public APIs, obviously. We could make the flag independent, or we could throw an exception when an invalid combination of flags is produced.

**Comment by jcouv** [comment]
Dankeshon

I'm re-running the integration branch. It has been consistently flaky in multiple PRs for me recently. Has anyone else run into this?
